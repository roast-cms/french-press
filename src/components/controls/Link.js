import LinkFilter from "@roast-cms/react-link-filter"
import React from "react"

/**
 * An implementation of `@roast-cms/react-link-filter` component that's built on top of React Router 4 with additional features, such as correcting user input (if they forgot to type http://), transforming absolute URLs on your domain to relative ones, and a proper function within React application that works in React Router 4 context.
 * @module Link
 */

export default props => {
  return <LinkFilter {...props}>{props.children}</LinkFilter>
}
