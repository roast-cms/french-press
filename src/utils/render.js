import React from "react"

import Unquote from "../components/controls/Unquote"

/**
 * Defines how all block-level nodes within the document are going to be rendered.
 * @function renderNode
 * @param props
 * @return {Object}
 */
export const renderNode = props => {
  const {node, attributes, children, isSelected, editor} = props
  const focus = editor.value.isFocused && isSelected
  const className = focus ? "focus" : "nofocus"

  const {components} = (editor && editor.props) || {}
  const {data} = node

  switch (node.type) {
    case "paragraph":
      return <p {...attributes}>{children}</p>
    case "heading":
      return <h3>{children}</h3>
    case "divider":
      return <hr className={className} />
    case "quote":
      return (
        <div style={{clear: "both"}}>
          {!props.readOnly && focus && (
            <Unquote
              className="fpe-unquote"
              contentEditable="false"
              spellCheck="false"
              suppressContentEditableWarning
              onClick={event => {
                event.preventDefault()
                editor.onChange(
                  editor.value
                    .change()
                    .setNodeByKey(attributes["data-key"], {
                      type: "paragraph",
                    })
                    .focus()
                )
              }}
              branded
            >
              Unquote
            </Unquote>
          )}
          <blockquote {...attributes} className={className}>
            {children}
          </blockquote>
        </div>
      )
    case "image": {
      if (components && components.Picture) {
        const Picture = components.Picture
        return <Picture {...props} />
      } else {
        // eslint-disable-next-line
        console.warn("<Picture /> component required to render images")
        return null
      }
    }
    case "link": {
      const href = data.get("href")

      if (components && components.Link) {
        const Link = components.Link
        return (
          <>
            <Link {...attributes} to={href}>
              {children}
            </Link>
            {" "}
          </>
        )
      } else {
        return (
          <>
            <a {...attributes} href={href}>
              {children}
            </a>
            {" "}
          </>
        )
      }
    }
    default:
      return <p {...attributes}>{children}</p>
  }
}

/**
 * Marks are inline "rules" for text that apply bold and italic formatting.
 * @function renderMark
 * @return {Object}
 */
export const renderMark = props => {
  const {children, mark} = props
  switch (mark.type) {
    case "bold":
      return <strong>{children}</strong>
    case "italic":
      return <em>{children}</em>
    default:
      return {children}
  }
}
