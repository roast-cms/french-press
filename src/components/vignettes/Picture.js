import React from "react"
import localForage from "localforage"
import styled from "styled-components"

import {base64ToBlob} from "../../utils/image"

export const Figure = styled.figure`
  margin: ${props => props.theme.size.block.spacing}em 0;
  &.focus {
    box-shadow: 0 0 0 4px ${props => props.theme.color.brand()};
  }
  img {
    display: block;
    max-width: 100%;
  }
`

/**
 * This component will render an image if it's an URL or request `data-uri` from browser's database using `localForage`.
 * @module Picture
 * @prop {Object} node Slate Editor node object.
 * @prop {Object} attributes Attributes for the Slate Editor node.
 * @prop {Boolean} isSelected Specifies whether the image is in user's focus.
 * @prop {Object} editor Slate Editor state.
 */
export default class extends React.PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      src: props.node.data.get("src") || "",
      key: "",
    }
  }
  componentDidMount = () => {
    const {node} = this.props
    const {data} = node
    const key = data.get("key")
    this.loadImage(data.get("file"), key, data.get("src"))
    this.setState({key})
  }
  componentWillUnmount = () => {
    this.state.src.includes("blob:") && URL.revokeObjectURL(this.state.src)
  }

  /**
   * Load image from URL or browser database via localForage
   * @function loadImage
   * @parameter {Object} file image file
   * @parameter {String} key key to image file in browser DB
   * @parameter {String} src
   */
  loadImage = (file, key, src) => {
    if (!key) {
      this.setState({src})
    } else {
      localForage.getItem(key).then(data => {
        if (data) {
          const src = URL.createObjectURL(base64ToBlob(data))
          this.setState({src})
        } else if (file && file.constructor !== Object) {
          const src = URL.createObjectURL(file)
          this.setState({src})
        }
      })
      this.setState({key})
    }
  }
  render = () => {
    const {isSelected, editor} = this.props
    const {src} = this.state

    const focus = editor.value.isFocused && isSelected
    const className = focus ? "focus" : "nofocus"

    return (
      <Figure className={className}>
        <img src={src} />
      </Figure>
    )
  }
}
