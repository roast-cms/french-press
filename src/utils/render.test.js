import React from "react"

import {shallow} from "enzyme"

import {renderNode} from "./render"
import Link from "../components/controls/Link"
import Picture from "../components/vignettes/Picture"

test("renderNode produces paragraph element.", () => {
  const element = (
    <p>
      <span>a</span>
    </p>
  )
  expect(
    renderNode({
      children: <span>a</span>,
      node: {type: "paragraph"},
      editor: {value: {isFocused: false}},
    })
  ).toEqual(element)
})

test("renderNode produces heading element.", () => {
  const element = <h3>a</h3>
  expect(
    renderNode({
      children: "a",
      node: {type: "heading"},
      editor: {value: {isFocused: false}},
    })
  ).toEqual(element)
})

test("renderNode produces divider element with correct class name.", () => {
  const element = <hr className="nofocus" />
  expect(
    renderNode({
      node: {type: "divider"},
      editor: {value: {isFocused: false}},
    })
  ).toEqual(element)
})

test("renderNode produces quote element that matches a snapshot for readOnly prop.", () => {
  const element = shallow(
    renderNode({
      children: "a",
      node: {type: "quote"},
      editor: {value: {isFocused: false}},
      readOnly: true,
    })
  )
  expect(element).toMatchSnapshot()
})
test("renderNode produces quote element that matches a snapshot for readOnly===false prop.", () => {
  const element = shallow(
    renderNode({
      children: "a",
      node: {type: "quote"},
      editor: {value: {isFocused: false}},
    })
  )
  expect(element).toMatchSnapshot()
})

test("renderNode produces picture element that matches a snapshot.", () => {
  const element = shallow(
    renderNode({
      node: {type: "picture"},
      editor: {value: {isFocused: false}, props: {components: {Picture}}},
    })
  )
  expect(element).toMatchSnapshot()
})

test("renderNode produces link element with correct href prop.", () => {
  const element = (
    <Link to="https://github.com/roast-cms/french-press-editor">a</Link>
  )
  expect(
    renderNode({
      node: {
        type: "link",
        data: {
          get: () => "https://github.com/roast-cms/french-press-editor",
        },
      },
      editor: {
        value: {},
        props: {
          options: {
            domain: "https://github.com",
          },
        },
      },
      children: "a",
    })
  ).toEqual(element)
})
