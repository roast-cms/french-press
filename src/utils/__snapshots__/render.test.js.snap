// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderNode produces picture element that matches a snapshot. 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <p />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": undefined,
    },
    "ref": null,
    "rendered": null,
    "type": "p",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": undefined,
      },
      "ref": null,
      "rendered": null,
      "type": "p",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`renderNode produces quote element that matches a snapshot for readOnly prop. 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <div
    style={
      Object {
        "clear": "both",
      }
    }
  >
    <blockquote
      className="nofocus"
    >
      a
    </blockquote>
  </div>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        false,
        <blockquote
          className="nofocus"
        >
          a
        </blockquote>,
      ],
      "style": Object {
        "clear": "both",
      },
    },
    "ref": null,
    "rendered": Array [
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "a",
          "className": "nofocus",
        },
        "ref": null,
        "rendered": "a",
        "type": "blockquote",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          false,
          <blockquote
            className="nofocus"
          >
            a
          </blockquote>,
        ],
        "style": Object {
          "clear": "both",
        },
      },
      "ref": null,
      "rendered": Array [
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "a",
            "className": "nofocus",
          },
          "ref": null,
          "rendered": "a",
          "type": "blockquote",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`renderNode produces quote element that matches a snapshot for readOnly===false prop. 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <div
    style={
      Object {
        "clear": "both",
      }
    }
  >
    <blockquote
      className="nofocus"
    >
      a
    </blockquote>
  </div>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [

        <blockquote
          className="nofocus"
        >
          a
        </blockquote>,
      ],
      "style": Object {
        "clear": "both",
      },
    },
    "ref": null,
    "rendered": Array [

      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "a",
          "className": "nofocus",
        },
        "ref": null,
        "rendered": "a",
        "type": "blockquote",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [

          <blockquote
            className="nofocus"
          >
            a
          </blockquote>,
        ],
        "style": Object {
          "clear": "both",
        },
      },
      "ref": null,
      "rendered": Array [

        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "a",
            "className": "nofocus",
          },
          "ref": null,
          "rendered": "a",
          "type": "blockquote",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
